libjcat/jcat-engine-pkcs7.c

```diff
@@ -0,0 +1,834 @@
+	g_autofree gchar *filename = NULL;
+	filename = g_build_filename (jcat_engine_get_localstatedir (engine),
+				     "pki", "client.pem", NULL);
+	return jcat_engine_pkcs7_load_crt_from_filename (filename,
+							GNUTLS_X509_FMT_PEM,
+							error);
+}
+
+static gnutls_pubkey_t
+jcat_engine_pkcs7_load_pubkey_from_privkey (gnutls_privkey_t privkey, GError **error)
+{
+	g_auto(gnutls_pubkey_t) pubkey = NULL;
+	int rc;
+
+	/* get the public key part of the private key */
+	rc = gnutls_pubkey_init (&pubkey);
+	if (rc < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "pubkey_init: %s [%i]",
+			     gnutls_strerror (rc), rc);
+		return NULL;
+	}
+	rc = gnutls_pubkey_import_privkey (pubkey, privkey, 0, 0);
+	if (rc < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "pubkey_import_privkey: %s [%i]",
+			     gnutls_strerror (rc), rc);
+		return NULL;
+	}
+
+	/* success */
+	return g_steal_pointer (&pubkey);
+}
+
+/* generates a private key just like:
+ *  `certtool --generate-privkey` */
+static gboolean
+jcat_engine_pkcs7_ensure_private_key (JcatEnginePkcs7 *self, GError **error)
+{
+#ifdef HAVE_GNUTLS_3_6_0
+	JcatEngine *engine = JCAT_ENGINE (self);
+	gnutls_datum_t d = { 0 };
+	int bits;
+	int key_type = GNUTLS_PK_RSA;
+	int rc;
+	g_autofree gchar *fn = NULL;
+	g_auto(gnutls_x509_privkey_t) key = NULL;
+	g_auto(gnutls_x509_spki_t) spki = NULL;
+	g_autoptr(GFile) file = NULL;
+	g_autoptr(gnutls_data_t) d_payload = NULL;
+
+	/* check exists */
+	fn = g_build_filename (jcat_engine_get_localstatedir (engine),
+			       "pki", "secret.key", NULL);
+	if (g_file_test (fn, G_FILE_TEST_EXISTS))
+		return TRUE;
+
+	/* initialize key and SPKI */
+	rc = gnutls_x509_privkey_init (&key);
+	if (rc < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "privkey_init: %s [%i]",
+			     gnutls_strerror (rc), rc);
+		return FALSE;
+	}
+	rc = gnutls_x509_spki_init (&spki);
+	if (rc < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "spki_init: %s [%i]",
+			     gnutls_strerror (rc), rc);
+		return FALSE;
+	}
+
+	/* generate key */
+	bits = gnutls_sec_param_to_pk_bits (key_type, GNUTLS_SEC_PARAM_HIGH);
+	g_debug ("generating a %d bit %s private key...",
+		 bits, gnutls_pk_algorithm_get_name (key_type));
+	rc = gnutls_x509_privkey_generate2(key, key_type, bits, 0, NULL, 0);
+	if (rc < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "privkey_generate2: %s [%i]",
+			     gnutls_strerror (rc), rc);
+		return FALSE;
+	}
+	rc = gnutls_x509_privkey_verify_params (key);
+	if (rc < 0) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "privkey_verify_params: %s [%i]",
+			     gnutls_strerror (rc), rc);
+		return FALSE;
+	}
+
+	/* create parents if required */
+	if (!jcat_mkdir_parent (fn, error))
+		return FALSE;
```

Comment:
I would think make the directory first, then the key, then export it.  That way if the directory can't be made you fail earlier.

Reply:
Agree, fixed.